synapse/storage/databases/main/event_federation.py

```diff
@@ -714,77 +722,152 @@ def _get_auth_chain_difference_txn(
             sql = """
-                SELECT b.event_id, MAX(e.depth) FROM events as e
+                SELECT backward_extrem.event_id, event.depth FROM events as event
                 /**
                  * Get the edge connections from the event_edges table
                  * so we can see whether this event's prev_events points
                  * to a backward extremity in the next join.
                  */
-                INNER JOIN event_edges as g
-                ON g.event_id = e.event_id
+                INNER JOIN event_edges as edge
+                ON edge.event_id = event.event_id
                 /**
                  * We find the "oldest" events in the room by looking for
                  * events connected to backwards extremeties (oldest events
                  * in the room that we know of so far).
                  */
-                INNER JOIN event_backward_extremities as b
-                ON g.prev_event_id = b.event_id
-                WHERE b.room_id = ? AND g.is_state is ?
-                GROUP BY b.event_id
+                INNER JOIN event_backward_extremities as backward_extrem
+                ON edge.prev_event_id = backward_extrem.event_id
+                /**
+                 * We use this info to make sure we don't retry to use a backfill point
+                 * if we've already attempted to backfill from it recently.
+                 */
+                LEFT JOIN event_backfill_attempts as backfill_attempt_info
+                ON backfill_attempt_info.event_id = backward_extrem.event_id
+                WHERE
+                    backward_extrem.room_id = ?
+                    /* We only care about normal events because TODO: why */
+                    AND edge.is_state is ? /* False */
+                    /**
+                     * We only want backwards extremities that are older than or at
+                     * the same position of the given `current_depth` (where older
+                     * means less than the given depth) because we're looking backwards
+                     * from the `current_depth` when backfilling.
+                     *
+                     *                           current_depth (ignore events that come after this, ignore 2-4)
+                     *                           |
+                     * <oldest-in-time> [0]<--[1]▼<--[2]<--[3]<--[4] <newest-in-time>
+                     */
+                    AND event.depth <= ? /* current_depth */
+                    /**
+                     * Exponential back-off (up to the upper bound) so we don't retry the
+                     * same backfill point over and over. ex. 2hr, 4hr, 8hr, 16hr, etc
+                     */
+                    AND (
+                        backfill_attempt_info IS NULL
+                        OR ? /* current_time */ >= backfill_attempt_info.last_attempt_ts + least(2^backfill_attempt_info.num_attempts * ? /* step */, ? /* upper bound */)
```

Comment:
One problem with doing the calculation in SQL is that SQLite doesn't like the `^` power operator, `sqlite3.OperationalError: unrecognized token: "^"`

Reply:
via https://stackoverflow.com/a/67105253/796832, it looks like SQLite does have `power(X, Y)` available but it was added in [SQLite v3.35.0](https://www.sqlite.org/changes.html#version_3_35_0) (2021-03-12) and only as a compile-time option  `-DSQLITE_ENABLE_MATH_FUNCTIONS` 😩

Reply:
via https://stackoverflow.com/a/55264405/796832, it's possible to create a user-defined function from Python and looks like we already do it so this might be viable:

https://github.com/matrix-org/synapse/blob/c4e29b6908ac8ae57b5e9a3e7662ad638b61e94a/synapse/storage/engines/sqlite.py#L98

Reply:
You can use the left shift operator: `2 << (n - 1)` instead of `pow(2, n)`. It's in Debian oldstable's version of SQLite, so it should be supported. :)

Reply:
Ahh, great thinking @reivilibre! 😇

For onlookers, only works with powers of `2` because left shift is a binary operation aka working with base-2.

Added in SQLite 2.0.3 (20001-10-13), https://www.sqlite.org/changes.html#version_2_0_3